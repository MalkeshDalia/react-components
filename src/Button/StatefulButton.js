import React from 'react';
import pick from 'lodash.pick';
import { skinnable, props, t } from '../utils';
import Button, { ButtonPropTypes } from './Button';

// const PromiseType = t.irreducible('Promise', x => x instanceof Promise);

/**
 * A generic button wrapper; it handles processing/success/failure states
 *
 * If buttonState is passed in as a prop, all internal states are overridden.
 *
 * stableSuccess should not be changed for the lifetime of the component
 *
 * Otherwise, the following happens:
 *
 *   internalState = null
 *     |
 *     ? click
 *     | --> invoke onClick and grab promise
 *     V
 *   internalState = processing
 *     |                    \
 *     ? promise success     ------? promise failure
 *     |                            \
 *     V                             V
 *   internalState = success       internalState = error
 *     |                 \                           \
 *     |                  --------                    --------------------
 *     |                          \                                       \
 *     ? stableSuccess is true     ? stableSuccess is false                |
 *     |                           |                                       |
 *     O wait for prop change      * < wait for timerMillis milliseconds > *
 *                                 |                                      /
 *                                 |              ------------------------
 *                                 |             /
 *                                 V             V
 *                               internalState = null
 *
 */

 /**
 * ready or not-allowed; use it if you want button to handle its internal state and onClick is a promise
 */

@skinnable()
@props({
  ...ButtonPropTypes,
  baseState: t.maybe(t.enums.of(['ready', 'success', 'not-allowed'])),
  stableSuccess: t.maybe(t.Boolean),
  timerMillis: t.maybe(t.Number)
})
export default class StatefulButton extends React.PureComponent {

  static defaultProps = {
    stableSuccess: false,
    timerMillis: 2000
  };

  constructor(props) {
    super(props);
    this.timeoutId = null;
    this.resetInternalStateAfterProcessing = false;
    this.state = {
      internalState: null
    };
  }

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  doResetInternalState = () => {
    if (this._isMounted) {
      this.setState({
        internalState: null
      });
    }
  }

  componentWillReceiveProps(props) {
    if (process.env.NODE_ENV !== 'production') {
      t.assert(props.stableSuccess === this.props.stableSuccess);
    }

    if (props.buttonState) {
      if (this.timeoutId) {
        clearTimeout(this.timeoutId);
        this.timeoutId = null;
      }
    }
    if (props.baseState !== this.props.baseState) {
      if (this.state.internalState === 'processing') {
        this.resetInternalStateAfterProcessing = true;
      } else {
        this.doResetInternalState();
      }
    }
  }

  doResetInternalStateAfterTimer = () => {
    this.timeoutId = setTimeout(() => {
      this.timeoutId = null;
      this.doResetInternalState();
    }, this.props.timerMillis);
  }

  attachPromiseHandlers = (promise) => {
    promise.then(() => {
      return this._isMounted && this.setState({
        internalState: 'success'
      }, () => {
        if (!this.props.stableSuccess) {
          this.doResetInternalStateAfterTimer();
        } else if (this.resetInternalStateAfterProcessing) {
          this.doResetInternalState();
        }
      });
    }).catch(() => {
      if (this._isMounted) {
        this.setState({
          internalState: 'error'
        }, () => this.doResetInternalStateAfterTimer() );
      }
    });
  }

  getButtonState = () => (this.props.buttonState || this.state.internalState || this.props.baseState);

  onClick = () => {
    if (this.getButtonState() === 'ready') {
      this.resetInternalStateAfterProcessing = false;
      const promise = this.props.onClick();
      if (!this.props.buttonState) {
        this.setState({
          internalState: 'processing'
        }, () => this.attachPromiseHandlers(promise) );
      }
    }
  };

  getLocals() {
    const buttonProps = pick(this.props, Object.keys(ButtonPropTypes));
    return {
      ...buttonProps,
      onClick: this.onClick,
      buttonState: this.getButtonState()
    };
  }

  template(props) {
    return <Button {...props} />;
  }

}
