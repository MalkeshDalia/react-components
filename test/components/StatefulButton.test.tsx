import * as React from 'react';
import { mount } from 'enzyme';

import StatefulButton, { StatefulButtonProps } from '../../src/Button/StatefulButton';
import clone = require('lodash/clone');

let consoleError: jest.SpyInstance<{}>

beforeAll(() => {
  consoleError = jest.spyOn(console, 'error');
});

afterEach(() => {
  expect(consoleError).not.toHaveBeenCalled();
});

function timeoutPromise(millis) {
  return new Promise((resolve) => {
    setTimeout(resolve, millis);
  });
}

const exampleProps: StatefulButtonProps = {
  ...StatefulButton.defaultProps,
  label: 'BeautifulButton',
  onClick: timeoutPromise.bind(this, 5),
  baseState: 'ready',
  stableSuccess: false,
  timerMillis: 5
};

type Overrides = {
  children?: string,
  [x: string]: any
};

function jsxComponent({ children, ...overrides } : Overrides) {
  return <StatefulButton {...exampleProps}  label={children || 'BeautifulButton'} {...overrides} />;
}

describe('StatefulButton', () => {

  it('renders correctly', () => {
    const component = mount(
      <StatefulButton {...exampleProps} />
    );
    expect(component.html()).toMatchSnapshot();
  });

  it('behaves correctly in base state', () => {
    const component = mount(jsxComponent({
      stableSuccess: false
    }));
    expect(component.prop('baseState')).toBe('ready');
    expect(component.find('.button-label span').first().text()).toBe('BeautifulButton');
  });

  it('passes through state if provided as prop', () => {
    const component = mount(jsxComponent({
      children: 'Button with error',
      buttonState: 'error'
    }));
    const buttonInner = component.find('.button-inner');
    expect(buttonInner.hasClass('is-error')).toBe(true);
    expect(buttonInner.children()).toBeDefined();
    expect(component.prop('label')).toBe('Button with error');
  });

  describe('with unstable success state', () => {
    it('switches to processing, success, ready on click', () => {
      const component = mount(jsxComponent({
        stableSuccess: false
      }));
      const buttonInner = component.find('.button-inner');
      buttonInner.simulate('click');
      expect(buttonInner.hasClass('is-processing')).toBe(true);
      return timeoutPromise(6).then(() => {
        expect(buttonInner.hasClass('is-success')).toBe(true);
        return timeoutPromise(5);
      }).then(() =>
        expect(buttonInner.hasClass('is-ready')).toBe(true)
      );
    });
  });

  describe('with stable success state', () => {
    it('switches to processing, success and stop there, on click', () => {
      const component = mount(jsxComponent({
        stableSuccess: true
      }));

      const buttonInner = component.find('.button-inner');
      buttonInner.simulate('click');
      expect(buttonInner.hasClass(`is-processing`)).toBe(true);

      return timeoutPromise(6).then(() => {
        expect(buttonInner.hasClass('is-success')).toBe(true);
        return timeoutPromise(5);
      }).then(() =>
        expect(buttonInner.hasClass('is-success')).toBe(true)
      );
    });

    it('switches to processing, success and back to baseState when baseState is changed after click', () => {
      const component = mount(jsxComponent({
        stableSuccess: true
      }));
      const buttonInner = component.find('.button-inner');
      buttonInner.simulate('click');
      expect(buttonInner.hasClass('is-processing')).toBe(true);
      const newProps = clone(component.props);
      newProps.baseState = 'not-allowed';
      component.setProps(newProps);
      return timeoutPromise(7).then(() =>
        expect(buttonInner.hasClass('is-not-allowed')).toBe(true)
      );
    });
  });

});
