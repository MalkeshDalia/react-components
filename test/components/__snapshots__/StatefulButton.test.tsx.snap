// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatefulButton renders correctly 1`] = `"<div class=\\"button\\" style=\\"display: inline-block;\\"><div class=\\"react-flex-view align-content-center justify-content-center button-inner is-default is-medium is-ready\\"><!-- react-text: 3 --><!-- /react-text --><div class=\\"react-flex-view flex-column justify-content-center align-content-center button-label\\"><div style=\\"width: 100%;\\"><div><span style=\\"display: block; white-space: nowrap; width: 100%; overflow: hidden; text-overflow: ellipsis; position: relative;\\">BeautifulButton<div class=\\"resize-sensor\\" style=\\"position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;\\"><div class=\\"resize-sensor-expand\\" style=\\"position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: hidden; z-index: -1; visibility: hidden;\\"><div style=\\"position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;\\"></div></div><div class=\\"resize-sensor-shrink\\" style=\\"position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: hidden; z-index: -1; visibility: hidden;\\"><div style=\\"position: absolute; left: 0; top: 0; transition: 0s; width: 200%; height: 200%\\"></div></div></div></span><span style=\\"position: fixed; visibility: hidden; pointer-events: none; top: 0px; left: 0px;\\">BeautifulButton</span></div></div></div></div></div>"`;
